### PR TITLE
filed: detect integer overflow during backup

### PR DESCRIPTION
**Pull request by pstorz**
The filedaemon now makes sure that the fileindex does not overflow during the backup.  The upper limit for file indices is currently 2^(31) - 1, but we now also build a test file daemon that has this limit set to 20, so that we can actually test the overflow detection.

#### Please check

- [x] Short description and the purpose of this PR is present _above this paragraph_
- [x] Your name is present in the AUTHORS file (optional)

If you have any questions or problems, please give a comment in the PR.

### Helpful documentation and best practices

- [Git Workflow](https://docs.bareos.org/DeveloperGuide/gitworkflow.html)
- [Automatic Sourcecode Formatting](https://docs.bareos.org/DeveloperGuide/generaldevel.html#automatic-sourcecode-formatting)
- [Check your commit messages](https://docs.bareos.org/DeveloperGuide/gitworkflow.html#commits)
- [Boy Scout Rule](https://docs.bareos.org/DeveloperGuide/generaldevel.html#boy-scout-rule)

### Checklist for the _reviewer_ of the PR (will be processed by the Bareos team)
Make sure you check/merge the PR using `devtools/pr-tool` to have some simple automated checks run and a proper changelog record added.

##### General
- [x] Is the PR title usable as CHANGELOG entry?
- [x] Purpose of the PR is understood
- [x] Commit descriptions are understandable and well formatted
- [x] Required backport PRs have been created : #2024 
- [x] Correct milestone is set

##### Source code quality
- [x] Source code changes are understandable
- [x] Variable and function names are meaningful
- [x] Code comments are correct (logically and spelling)
- [x] Required documentation changes are present and part of the PR

##### Tests
- [x] Decision taken that a test is required (if not, then remove this paragraph)
- [x] The choice of the type of test (unit test or systemtest) is reasonable
- [x] Testname matches exactly what is being tested
- [x] On a fail, output of the test leads quickly to the origin of the fault

### CODE REVIEW

File: core/src/filed/backup.cc [hunk 1]
```diff
@@ -1604,7 +1604,12 @@ bool EncodeAndSendAttributes(JobControlRecord* jcr,
 
   {
     std::unique_lock l(jcr->mutex_guard());
-    jcr->JobFiles++;                   /* increment number of files sent */
+    int IncreasedJobFiles;
+    if (__builtin_sadd_overflow(jcr->JobFiles, 1, &IncreasedJobFiles)) {
+      Jmsg0(jcr, M_FATAL, 0, T_("jcr->JobFiles overflow prevented!\n"));
+      return false;
+    }
+    jcr->JobFiles = IncreasedJobFiles; /* increment number of files sent */
```
Comment (sebsura) [inline core/src/filed/backup.cc:1612]:
```suggestion
    if (jcr->JobFiles == std::numeric_limits<decltype(jcr->JobFiles)>::max()) {
      Jmsg0(jcr, M_FATAL, 0, T_("You are trying to back up too many files at once.  Please split up your fileset.\n"));
      return false;
    }
    jcr->JobFiles += 1; /* increment number of files sent */
```

Comment (sebsura) [reply core/src/filed/backup.cc:1612]: `__builtin_sadd_overflow` is a gcc builtin and as such not available on windows.   `jcr->JobFiles` is also unsigned, why are we using a simple int here ? If we want to ensure that there is no overflow even with a signed value we should use something like
```
static_cast<std::make_signed_t<decltype(jcr->JobFiles)>>(jcr->JobFiles) >= 0
```
after the increment.

**Review by sebsura — CHANGES_REQUESTED** [review]

File: core/src/filed/backup.cc (hunk 1 above)
Comment (arogge) [reply core/src/filed/backup.cc:1612]: I'd even go so far to use a macro like this
```
#if not defined JOB_FILES_MAX
#define JOB_FILES_MAX std::numeric_limits<decltype(JobControlRecord::JobFiles)::max())
#endif
```
Because then we could build with `-DJOB_FILES_MAX=20` and run a proper test.

**Review by sebsura — CHANGES_REQUESTED** [review]
We should probably add checks to `::SendBaseFileList` and `::SendDeletedList`; either they should check for the return value of `EncodeAndSendAttributes` or they should check if the job was canceled and stop.

Maybe even both.

File: core/src/filed/backup.cc [hunk 2]
```diff
@@ -1605,6 +1605,13 @@ bool EncodeAndSendAttributes(JobControlRecord* jcr,
   {
     std::unique_lock l(jcr->mutex_guard());
     jcr->JobFiles++;                   /* increment number of files sent */
+    if (jcr->JobFiles > MAXIMUM_ALLOWED_FILES_PER_JOB) {
+      Jmsg1(jcr, M_FATAL, 0,
+            "FileIndex overflow detected. Please split the fileset to backup "
+            "less files.\n");
+      return false;
+    }
+
```
Comment (sebsura) [inline core/src/filed/backup.cc:1614]: This does not work.  You cannot be bigger than the max.

The correct way to check for this is
```
if (jcr->JobFiles == MAXIMUM_ALLOWED_FILES_PER_JOB) { ... }
jcr->JobFiles += 1;
```
Currently `jcr->JobFiles` will overflow causing the check to always be false.

File: core/src/filed/backup.h [hunk 3]
```diff
@@ -72,6 +72,10 @@ void StripPath(FindFilesPacket* ff_pkt);
 void UnstripPath(FindFilesPacket* ff_pkt);
 int SaveFile(JobControlRecord* jcr, FindFilesPacket* ff_pkt, bool top_level);
 
+// allow to set the allowed maximum to something lower for testing purposes
+#ifndef MAXIMUM_ALLOWED_FILES_PER_JOB
+#define MAXIMUM_ALLOWED_FILES_PER_JOB std::numeric_limits<decltype(JobControlRecord::JobFiles)>::max()
+#endif
```
Comment (sebsura) [inline core/src/filed/backup.h:78]: This is fine for now, but we should think of a way to do this in which it is not only testable but also _tested_.

Maybe for 25 we can change this to be a run time value that you can set when starting the FD.  This way we can actually set up an automated test for it!

**Review by sebsura — CHANGES_REQUESTED** [review]

File: core/src/filed/backup.cc [hunk 4]
```diff
@@ -1602,11 +1613,15 @@ bool EncodeAndSendAttributes(JobControlRecord* jcr,
...
-    std::unique_lock l(jcr->mutex_guard());
-    jcr->JobFiles++;                   /* increment number of files sent */
-    ff_pkt->FileIndex = jcr->JobFiles; /* return FileIndex */
+  if (std::optional findex = get_next_findex(jcr)) {
+    ff_pkt->FileIndex = *findex;
     PmStrcpy(jcr->fd_impl->last_fname, ff_pkt->fname);
+  } else {
+    Jmsg1(jcr, M_FATAL, 0,
+          "FileIndex overflow detected. Please split the fileset to backup "
+          "less files.\n");
```
Comment (sebsura) [inline core/src/filed/backup.cc:1622]:
```suggestion
          "fewer files.\n");
```

File: core/src/filed/backup.cc [hunk 5]
```diff
@@ -1602,11 +1613,15 @@ bool EncodeAndSendAttributes(JobControlRecord* jcr,
...
 
-  {
-    std::unique_lock l(jcr->mutex_guard());
-    jcr->JobFiles++;                   /* increment number of files sent */
-    ff_pkt->FileIndex = jcr->JobFiles; /* return FileIndex */
+  if (std::optional findex = get_next_findex(jcr)) {
+    ff_pkt->FileIndex = *findex;
     PmStrcpy(jcr->fd_impl->last_fname, ff_pkt->fname);
+  } else {
+    Jmsg1(jcr, M_FATAL, 0,
```
Comment (sebsura) [inline core/src/filed/backup.cc:1620]: Currently this causes the backup to end in state `*** Backup Error ***`.  Should we maybe set the state to `JS_Incomplete` instead ?  Then the end state will show as `Backup failed -- incomplete`

Im not sure which one is better, what do you think @pstorz  ?

**Review by sebsura — CHANGES_REQUESTED** [review]

**Review by sebsura — APPROVED** [review]
